### ISSUE TITLE
Maps of scalar key/value with missing key/value cause key/value null check fails

### ISSUE THREAD

**Issue by zsoltvilagos** [square/wire#2231]
### TL;DR;

ScalaPB encodes `"some" -> ""` for a field of type `map<string, string>` differently than Wire and `protoc`. While decoding with `protos` works, Wire decode fails.

### Detailed

We have a proto schema with a `map<string, string>` field like https://github.com/square/wire/blob/4.3.0/wire-library/wire-tests/src/commonTest/shared/proto/proto3/map_types.proto or https://github.com/square/wire/blob/4.3.0/wire-protoc-compatibility-tests/src/main/proto/squareup/proto3/kotlin/map_types.proto.

We use this with multiple techs including protoc, ScalaPB and Wire. One of our Scala code produces such a proto message with a `"some" -> ""` mapping.

Decoding that message with protoc works but Wire fails with:
```
Map entry with null value
java.lang.IllegalStateException: Map entry with null value
	at com.squareup.wire.MapProtoAdapter.decode(ProtoAdapter.kt:517)
	at com.squareup.wire.MapProtoAdapter.decode(ProtoAdapter.kt:440)
	at squareup.proto3.kotlin.MapTypes$Companion$ADAPTER$1.decode(MapTypes.kt:446)
	at squareup.proto3.kotlin.MapTypes$Companion$ADAPTER$1.decode(MapTypes.kt:345)
	at com.squareup.wire.ProtoAdapter.decode(ProtoAdapter.kt:374)
	at com.squareup.wire.ProtoAdapter.decode(ProtoAdapter.kt:372)
```

Producing the same message with protoc or Wire and then decoding either with protoc or Wire has no issue.

I stepped through how Wire decodes the ScalaPB produced message and identified that the map entry has key tag but not value tag.

For such, https://github.com/square/wire/blob/4.3.0/wire-library/wire-runtime/src/commonMain/kotlin/com/squareup/wire/ProtoAdapter.kt#L490 leaves `var value` on null and then
`check(value != null) { "Map entry with null value" }` produces the above failure.

Checked protoc and Wire encoding, they produce both tags explicitly (value tag length 0).

Looked into finding clues in the official Proto Language Guide. From https://developers.google.com/protocol-buffers/docs/proto3#maps, the below seems applicable:
> If you provide a key but no value for a map field, the behavior when the field is serialized is language-dependent. In C++, Java, Kotlin, and Python the default value for the type is serialized, while in other languages nothing is serialized

Feels like ScalaPB is doing that "other languages" mode.

I also checked the `"" -> "some"` case, this time manually crafting proto wire bytes instead of by ScalaPB and decoding with protoc and Wire. Same behavior. Protoc works, Wire fails with matching exception like above.
And also https://github.com/square/wire/blob/4.3.0/wire-protoc-compatibility-tests/src/main/proto/squareup/proto3/kotlin/pizza/pizza.proto for `repeated string`. No issue.

Following CONTRIBUTING.md, I came up with a failing test and an idea to fix it which I would offer as a PR from a personal fork, but for that I need to check the CLA with my employer.
Update: checked, CLA filled & sent, opened https://github.com/square/wire/pull/2232

**Comment by swankjesse** [comment]
Can fix. Thanks for the detailed report.

**Comment by zsoltvilagos** [comment]
> Can fix. Thanks for the detailed report.

@swankjesse Just to be sure that it not missed because I just recently updated the description that I opened a PR: https://github.com/square/wire/pull/2232

**Comment by zsoltvilagos** [comment]
#2232 merged, closing issue.
